File: Difficulty: Medium/K Sum Paths/k-sum-paths.js
//{ Driver Code Starts
// Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

function buildTree(str) {
    // Corner Case
    if (str.length === 0 || str[0] === "N") return null;

    // Create the root of the tree
    let root = new Node(parseInt(str[0]));

    // Push the root to the queue
    let queue = [];
    let start = 0;
    queue.push(root);

    // Starting from the second element
    let i = 1;
    while (queue.length !== start && i < str.length) {

        // Get and remove the front of the queue
        let currNode = queue[start];
        start++;

        // Get the current node's value from the string
        let currVal = str[i];

        // If the left child is not null
        if (currVal !== "N") {

            // Create the left child for the current node
            currNode.left = new Node(parseInt(currVal));

            // Push it to the queue
            queue.push(currNode.left);
        }

        // For the right child
        i++;
        if (i >= str.length) break;
        currVal = str[i];

        // If the right child is not null
        if (currVal !== "N") {

            // Create the right child for the current node
            currNode.right = new Node(parseInt(currVal));

            // Push it to the queue
            queue.push(currNode.right);
        }
        i++;
    }

    return root;
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {

        let s_tree = readLine().trim().split(" ");
        let K = parseInt(readLine());
        let root = buildTree(s_tree);
        let obj = new Solution();
        let res = obj.sumK(root, K);
        console.log(res);
        console.log("~");
    }
}

// } Driver Code Ends


// User function Template for javascript

/*
class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}
*/

/**
 * @param {Node} root
 * @param {number} K
 * @return {number}
 */

class Solution {
    solve(root, K, count, path) {
        if (root == null) {
            return;
        }
        
        // Add the current node's data to the path
        path.push(root.data);
        
        // Explore left and right subtrees
        this.solve(root.left, K, count, path);
        this.solve(root.right, K, count, path);
        
        // Check if any subpath from this node to the root sums to K
        let sum = 0;
        for (let i = path.length - 1; i >= 0; i--) {
            sum += path[i];
            if (K == sum) {
                count[0]++; // Increment the count if sum equals K
            }
        }
        
        // Backtrack by removing the current node from the path
        path.pop();
    }

    sumK(root, K) {
        let count = [0];  // Array to keep track of the count
        let path = [];     // Array to store the current path
        this.solve(root, K, count, path);
        return count[0];   // Return the count of valid paths
    }
}

File: Difficulty: Medium/K Sum Paths/k-sum-paths.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};

// Function to Build Tree
Node *buildTree(string str) {
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return NULL;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;)
        ip.push_back(str);

    // Create the root of the tree
    Node *root = new Node(stoi(ip[0]));

    // Push the root to the queue
    queue<Node *> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size()) {

        // Get and remove the front of the queue
        Node *currNode = queue.front();
        queue.pop();

        // Get the current Node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N") {

            // Create the left child for the current Node
            currNode->left = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N") {

            // Create the right child for the current Node
            currNode->right = new Node(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}


// } Driver Code Ends

/*
struct Node {
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};
*/

class Solution {
  public:
    int ans = 0; // Variable to store the count of valid paths

    /**
     * Recursive function to find the number of paths in the tree where the sum of nodes equals k.
     * @param root - Pointer to the current node.
     * @param prefixSumMap - Hash map to store prefix sum frequencies.
     * @param k - Target sum.
     * @param currentSum - Sum of nodes from root to the current node.
     */
    void solve(Node *root, unordered_map<int, int> &prefixSumMap, int k, int currentSum) {
        if (!root) return; // Base case: If node is NULL, return.

        // Update the current sum by adding the current node's value
        currentSum += root->data;

        // Check if there exists a prefix sum such that currentSum - k exists in the map
        if (prefixSumMap.find(currentSum - k) != prefixSumMap.end()) {
            ans += prefixSumMap[currentSum - k]; // Add the count of such paths
        }

        // Store the current prefix sum in the map
        prefixSumMap[currentSum]++;

        // Recursively check left and right subtrees
        solve(root->left, prefixSumMap, k, currentSum);
        solve(root->right, prefixSumMap, k, currentSum);

        // Backtrack: Remove the current sum from the map before returning to the parent node
        prefixSumMap[currentSum]--; 
    }

    /**
     * Function to return the number of paths in the tree where the sum of nodes equals k.
     * @param root - Pointer to the root node of the tree.
     * @param k - Target sum.
     * @return Count of valid paths.
     */
    int sumK(Node *root, int k) {
        unordered_map<int, int> prefixSumMap; // Hash map to store prefix sum frequencies
        prefixSumMap[0] = 1; // Initialize with sum 0 to handle cases where path sum directly equals k
        solve(root, prefixSumMap, k, 0);
        return ans; // Return the count of valid paths
    }
};

//{ Driver Code Starts.

int main() {
    string tc;
    getline(cin, tc);
    int t = stoi(tc);

    while (t--) {
        string s, ch;
        getline(cin, s);
        Node *root = buildTree(s);

        string key;
        getline(cin, key);
        int k = stoi(key);
        Solution ob;
        cout << ob.sumK(root, k) << "\n";
        cout << "~\n";
    }
    return 0;
}
// } Driver Code Ends